charon/src/transform/remove_dynamic_checks.rs

```diff
@@ -1,346 +1,154 @@
-                                            let (dest, _) = s2.content.to_assign();
-                                            let meta0 = s0.meta;
-                                            let s0 = RawStatement::Assign(dest, op);
-                                            let s0 = Statement {
-                                                meta: meta0,
-                                                content: s0,
-                                            };
-                                            Statement {
-                                                meta: s2.meta,
-                                                content: RawStatement::Sequence(Box::new(s0), s3),
-                                            }
-                                        });
-                                        // A simplification happened
-                                        return true;
-                                    }
-                                }
-                            }
-                        }
-                    }
-                }
-            }
-        };
-
-        // No simplification
-        false
-    }
-}
+        // Overflow checks for signed division and remainder. They look like:
+        //   is_neg_1 := y == (-1)
+        //   is_min := x == INT::min
+        //   has_overflow := move (is_neg_1) & move (is_min)
+        //   assert(move has_overflow == false)
+        [rest @ .., Statement {
+            content: RawStatement::Assign(is_neg_1, Rvalue::BinaryOp(BinOp::Eq, _y_op, _minus_1)),
+            ..
+        }, Statement {
+            content: RawStatement::Assign(is_min, Rvalue::BinaryOp(BinOp::Eq, _x_op, _int_min)),
+            ..
+        }, Statement {
+            content:
+                RawStatement::Assign(
+                    has_overflow,
+                    Rvalue::BinaryOp(BinOp::BitAnd, Operand::Move(and_op1), Operand::Move(and_op2)),
+                ),
+            ..
+        }] if and_op1 == is_neg_1
+            && and_op2 == is_min
+            && cond == has_overflow
+            && *expected == false =>
+        {
+            rest
+        }
 
-impl<'tcx, 'ctx, 'a> MutAstVisitor for RemoveDynChecks<'tcx, 'ctx, 'a> {
-    fn spawn(&mut self, visitor: &mut dyn FnMut(&mut Self)) {
-        visitor(self)
-    }
+        // Overflow checks for left shift. They look like:
+        //   x := ...;
+        //   b := copy x < const 32; // or another constant
+        //   assert(move b == true);
+        [rest @ .., Statement {
+            content: RawStatement::Assign(x, _),
+            ..
+        }, Statement {
+            content:
+                RawStatement::Assign(
+                    has_overflow,
+                    Rvalue::BinaryOp(BinOp::Lt, Operand::Copy(lt_op2), Operand::Const(..)),
+                ),
+            ..
+        }] if lt_op2 == x && cond == has_overflow && *expected == true => rest,
```

Comment:
There is nothing for right shifts?

Reply:
Nope, they can't overflow

Reply:
And no one cares whether they underflow apparently

Reply:
I have to check that more thoroughly though, rust documentation isn't as clear as I'd like

Reply:
I was wrong: somehow I thought I had seen shl overflow checks and no shr overflow checks, but rust emits checks for both when not in release mode, and it's not about overflow. Since we use release mode we don't see them at all.